src/diff.rs

```diff
@@ -1,18 +1,45 @@
+// Copyright 2023 Developers of the reconcile project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! This [`module`](crate::diff) provides two traits:
+//! [`HashRangeQueryable`] and [`Diffable`].
+
 use std::ops::{Bound, RangeBounds};
 
 use serde::{Deserialize, Serialize};
 
+/// Provides the necessary methods to be able
+/// to efficiently determine and compare
+/// differences between two key stores:
+/// * [`hash`](HashRangeQueryable::hash),
+/// * [`insertion_position`](HashRangeQueryable::insertion_position),
+/// * [`key_at`](HashRangeQueryable::key_at),
+/// * [`len`](HashRangeQueryable::len)
+/// (with [`is_empty`](HashRangeQueryable::is_empty) as a default implementation).
+/// This is a low-level trait.
 pub trait HashRangeQueryable {
     type Key;
+    /// Hash of a given [`Key`](HashRangeQueryable::Key) range of type [`RangeBounds`].
     fn hash<R: RangeBounds<Self::Key>>(&self, range: &R) -> u64;
+    /// Insertion position of a given [`Key`](HashRangeQueryable::Key).
     fn insertion_position(&self, key: &Self::Key) -> usize;
+    /// Reference to the [`Key`](HashRangeQueryable::Key) at a given position.
```

Comment:
We cannot really enforce that in the signature, but that's what makes the most sense for the contract I think. That's consistent with e.g. the standard [`Index`](https://doc.rust-lang.org/std/ops/trait.Index.html#tymethod.index) trait.

Otherwise, we would need to change the return type to `Option<_>`. Maybe also the name to `get_key`.